File: app/src/main/java/com/kispoko/tome/sheet/widget/util/Position.java

package com.kispoko.tome.sheet.widget.util;


import android.widget.LinearLayout;

import com.kispoko.tome.exception.InvalidDataException;
import com.kispoko.tome.util.EnumUtils;
import com.kispoko.tome.lib.database.DatabaseException;
import com.kispoko.tome.lib.database.sql.SQLValue;
import com.kispoko.tome.lib.yaml.ToYaml;
import com.kispoko.tome.lib.yaml.YamlBuilder;
import com.kispoko.tome.lib.yaml.YamlParseException;
import com.kispoko.tome.lib.yaml.YamlParser;
import com.kispoko.tome.lib.yaml.error.InvalidEnumError;



/**
 * Position
 */
public enum Position implements ToYaml
{

    // VALUES
    // ------------------------------------------------------------------------------------------

    LEFT,
    TOP,
    RIGHT,
    BOTTOM;


    // CONSTRUCTORS
    // ------------------------------------------------------------------------------------------

    public static Position fromString(String positionString)
                  throws InvalidDataException
    {
        return EnumUtils.fromString(Position.class, positionString);
    }


    public static Position fromYaml(YamlParser yaml)
                  throws YamlParseException
    {
        String positionString = yaml.getString();
        try {
            return Position.fromString(positionString);
        } catch (InvalidDataException e) {
            throw YamlParseException.invalidEnum(new InvalidEnumError(positionString));
        }
    }


    public static Position fromSQLValue(SQLValue sqlValue)
                  throws DatabaseException
    {
        String enumString = "";
        try {
            enumString = sqlValue.getText();
            Position position = Position.fromString(enumString);
            return position;
        } catch (InvalidDataException e) {
            throw DatabaseException.invalidEnum(
                    new com.kispoko.tome.lib.database.error.InvalidEnumError(enumString));
        }
    }


    // TO YAML
    // ------------------------------------------------------------------------------------------

    /**
     * The Widget Content Alignment's yaml string representation.
     * @return The Yaml Builder.
     */
    public YamlBuilder toYaml()
    {
        return YamlBuilder.string(this.name().toLowerCase());
    }


    // LAYOUT ORIENTATION
    // ------------------------------------------------------------------------------------------

    public int linearLayoutOrientation()
    {
        switch (this)
        {
            case LEFT:
                return LinearLayout.HORIZONTAL;
            case TOP:
                return LinearLayout.VERTICAL;
            case RIGHT:
                return LinearLayout.HORIZONTAL;
            case BOTTOM:
                return LinearLayout.VERTICAL;
            default:
                return LinearLayout.HORIZONTAL;
        }
    }


}

File: app/src/main/java/com/kispoko/tome/model/sheet/style/Position.kt

package com.kispoko.tome.model.sheet.style


import android.widget.LinearLayout
import com.kispoko.tome.lib.orm.sql.SQLSerializable
import com.kispoko.tome.lib.orm.sql.SQLText
import com.kispoko.tome.lib.orm.sql.SQLValue
import effect.effError
import effect.effValue
import lulo.document.*
import lulo.value.UnexpectedType
import lulo.value.UnexpectedValue
import lulo.value.ValueError
import lulo.value.ValueParser
import java.io.Serializable



/**
 * Position
 */
sealed class Position : ToDocument, SQLSerializable, Serializable
{

    object Left : Position()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({"left"})

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("left")

    }


    object Top : Position()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({"top"})

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("top")

    }


    object Right : Position()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({"right"})

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("right")

    }


    object Bottom : Position()
    {
        // SQL SERIALIZABLE
        // -------------------------------------------------------------------------------------

        override fun asSQLValue() : SQLValue = SQLText({"bottom"})

        // TO DOCUMENT
        // -------------------------------------------------------------------------------------

        override fun toDocument() = DocText("bottom")

    }


    // -----------------------------------------------------------------------------------------
    // CONSTRUCTORS
    // -----------------------------------------------------------------------------------------

    companion object
    {
        fun fromDocument(doc : SchemaDoc) : ValueParser<Position> = when (doc)
        {
            is DocText -> when (doc.text)
            {
                "left"   -> effValue<ValueError,Position>(Position.Left)
                "top"    -> effValue<ValueError,Position>(Position.Top)
                "right"  -> effValue<ValueError,Position>(Position.Right)
                "bottom" -> effValue<ValueError,Position>(Position.Bottom)
                else     -> effError<ValueError,Position>(
                                    UnexpectedValue("Corners", doc.text, doc.path))
            }
            else       -> effError(UnexpectedType(DocType.TEXT, docType(doc), doc.path))
        }
    }


    fun isLeft() : Boolean = when (this)
    {
        is Left -> true
        else    -> false
    }


    fun isTop() : Boolean = when (this)
    {
        is Top -> true
        else   -> false
    }


    fun isRight() : Boolean = when (this)
    {
        is Right -> true
        else     -> false
    }


    fun isBottom() : Boolean = when (this)
    {
        is Bottom -> true
        else      -> false
    }


    fun linearLayoutOrientation() : Int = when (this)
    {
        is Top    -> LinearLayout.VERTICAL
        is Right  -> LinearLayout.HORIZONTAL
        is Bottom -> LinearLayout.VERTICAL
        is Left   -> LinearLayout.HORIZONTAL
    }

}



//
//
//    // LAYOUT ORIENTATION
//    // ------------------------------------------------------------------------------------------
//
//    public int linearLayoutOrientation()
//    {
//        switch (this)
//        {
//            case LEFT:
//                return LinearLayout.HORIZONTAL;
//            case TOP:
//                return LinearLayout.VERTICAL;
//            case RIGHT:
//                return LinearLayout.HORIZONTAL;
//            case BOTTOM:
//                return LinearLayout.VERTICAL;
//            default:
//                return LinearLayout.HORIZONTAL;
//        }
//    }

